### PR TITLE
add five new whitespace checks and fixes

### PR DESCRIPTION
**Pull request by rothsa**
adds checks for https://github.com/rodjek/puppet-lint/issues/651. 

These will create an exception minefield I'm afraid, however, I did try doing the checks isolated to resource and title at first, rather than behavior around braces and colons. Trying both ways, I think we're better off doing it this way because the spacing warnings here are generally the rule rather than the exception.

### CODE REVIEW

File: lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Public: Check the manifest tokens for resource type declaration that has no whitespace after a colon, and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb]: What statement specifically is this addressing? I think this is a good idea but don't see that in the guide.

File: lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+PuppetLint.new_check(:no_colon_whitespace_after) do
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb]: `whitespace_after_title_colon`?

File: lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  def check
+    tokens.each do |token|
+      next if token.next_token.nil?
+      unless [:WHITESPACE, :NEWLINE].include?(token.next_token.type) 
+        if (token.type == :COLON)
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb]: I think this might perform a little better with the `unless/if` order swapped, especially with large amounts of code to test.

I also think it needs to make sure the previous token is a title to prevent false positives. It's somewhat common to see condensed format JSON strings without spacing surrounding colons.

File: lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Public: Check the manifest tokens for resource type declaration that has no whitespace before a colon and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb]: That URL redirects to https://docs.puppet.com/puppet/latest/style_guide.html#spacing-indentation-and-whitespace, it's probably better to point directly at it.

File: lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Public: Check the manifest tokens for whitespace after the left bracket that is missing and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:right_lbrace_whitespace) do
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb]: `whitespace_before_title`?

File: lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  def check
+    tokens.each do |token|
+      next if token.next_token.nil?
+      if token.type == :LBRACE
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb]: Would it be better to look for title tokens and ensure the previous token is a single whitespace? `:LBRACE` can be found in many places, and `:WHITESPACE` could be more than a single space.

File: lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+PuppetLint.new_check(:colon_whitespace_before) do
+  def check
+    tokens.each do |token|
+      next if token.prev_token.nil?
+      if (token.type == :COLON && token.prev_token.type == :WHITESPACE)
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb]: Can you add a test for a colon following whitespace that isn't associated with a resource title, maybe in an exec's command string for instance?

File: lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Public: Check the manifest tokens for whitespace before left bracket that is missing and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:left_lbrace_whitespace) do
+  def check
+    tokens.each do |token|
+      unless token.next_token.nil?
+        if (token.type != :WHITESPACE && token.next_token.type == :LBRACE)
```
Comment (rnelson0) [inline lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb]: Will this catch a curly brace following text in a string, like JSON?

**Comment by rothsa** [comment]
Thank you very much for looking this over, I'll take care of those - is there a specific target I should shoot for with perf runtimes?

**Comment by rnelson0** [comment]
I don't know that there's a particular target, but I think `next_token_of(:WHITESPACE)` is generally better than iterating through all tokens, both from a performance and a maintainability perspective, for instance. I honestly don't have a codebase large enough for it to matter, maybe we can get someone at github or twitter to give it a run though.

File: lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb (hunk 1 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/no_colon_whitespace_after.rb]: I'll go ahead and remove this test, it addresses a spacing implied but not written out.

File: lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# Public: Check the manifest tokens for whitespace before left bracket that is missing and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:left_lbrace_whitespace) do
```
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb]: Is there any situation where the left curly brace should not have whitespace between it and the previous token? I didn't want to restrict it to resource types, only, but if we're better off being pedantic then I will do that.

File: lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb (hunk 8 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb]: Thanks, will make the test :)

File: lib/puppet-lint/plugins/check_whitespace/multiple_whitespace.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# Public: Check the manifest tokens for multiple spaces in whitespace that is not after a newline and record a warning for each instance found.
+# #
+# https://docs.puppet.com/guides/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:multiple_whitespace) do
```
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/multiple_whitespace.rb]: I'll see if I can make it work, and I agree, there are probably at least that and more exceptions. If I can't get a handle on it I'll restrict it to just whitespace around titles and resources, but I was hoping to make the check more general if possible.

File: lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb (hunk 5 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb]: Again here, I can't think of a situation where it would be preferable to have no whitespace after the left bracket, except for {} or a newline.

File: lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb (hunk 7 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/colon_whitespace_before.rb]: Done

File: lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb (hunk 8 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb]: done

File: lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb (hunk 6 above)
Comment (rothsa) [inline lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb]: @rnelson0 The reason I did not test for multiple here is because I found that including multiple whitespace in the test added a lot more control flow requirements to the fix logic than was palatable...additionally it made atomizing each actual problem longer and more complicated. That is why the multiple whitespace is its own test.

**Comment by rothsa** [comment]
Ok, tests are added. If you folks insist that these be isolated to titles and resources only, I will refactor to isolate the tests to just that.

**Comment by rnelson0** [comment]
Thanks for the updates. This looks awesome!

@rodjek I see no reason we could not add this to version 2.next, but I know the 3.x target is coming up soon. Merge as is, or wait for 3.x?

**Comment by mmerfort** [comment]
The new plugins show an error when you indent the parameter names on a class.
```Puppet
class test (
  String                       $foo = 'foo',
  Variant[String,Hash] $bar = 'bar'
) {
...
}
```
WARNING: too many spaces on line 2

I also had an unexpected warning with a create resource function:
```Puppet
  create_resource(
    test,
    $testhash,
    { require => File['/tmp/test'] }
  )
```
WARNING: space needed on left side of opening bracket on line <line with the require>

File: lib/puppet-lint/plugins/check_whitespace/multiple_whitespace.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Public: Check the manifest tokens for multiple spaces in whitespace that is not after a newline and record a warning for each instance found.
+# #
+# https://docs.puppet.com/puppet/latest/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:multiple_whitespace) do
```
Comment (baurmatt) [inline lib/puppet-lint/plugins/check_whitespace/multiple_whitespace.rb]: This is going to break this syntax:

```puppet
class test (
  Boolean                   $foo = false,
  Enum['ensure', 'present'] $bar = 'ensure',
) {
}
```

We prefer it, because we think it make the parameter/data type definition easier readable.

File: lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+# Public: Check the manifest tokens for whitespace before left bracket that is missing and record a warning for each instance found.
+# #
+# https://docs.puppet.com/puppet/latest/style_guide.html#spacing-indentation-and-whitespace 
+PuppetLint.new_check(:left_lbrace_whitespace) do
```
Comment (baurmatt) [inline lib/puppet-lint/plugins/check_whitespace/left_lbrace_whitespace.rb]: This throws errors:

```
mbaur(2018-10-30 14:17:08)  /tmp/puppet-lint  (spacing_warnings_fixes)
$ cat test.pp
#
class test (
  Hash $foo = {},
) {
  $test = merge({'bar' => 'baz'}, $foo )
}
mbaur(2018-10-30 14:17:45)  /tmp/puppet-lint  (spacing_warnings_fixes)
$ bundle exec puppet-lint test.pp
test.pp - ERROR: test not in autoload module layout on line 2
test.pp - WARNING: space needed on right side of opening bracket on line 5
test.pp - WARNING: space needed on left side of opening bracket on line 5
```

File: lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# Public: Check the manifest tokens for whitespace after the left bracket that is missing and record a warning for each instance found.
+# #
+# https://docs.puppet.com/puppet/latest/style_guide.html#spacing-indentation-and-whitespace
+PuppetLint.new_check(:right_lbrace_whitespace) do
```
Comment (baurmatt) [inline lib/puppet-lint/plugins/check_whitespace/right_lbrace_whitespace.rb]: This breaks the same way as left_lbrace_whitespace:

```
mbaur(2018-10-30 14:17:08)  /tmp/puppet-lint  (spacing_warnings_fixes)
$ cat test.pp
#
class test (
  Hash $foo = {},
) {
  $test = merge({'bar' => 'baz'}, $foo )
}
mbaur(2018-10-30 14:17:45)  /tmp/puppet-lint  (spacing_warnings_fixes)
$ bundle exec puppet-lint test.pp
test.pp - ERROR: test not in autoload module layout on line 2
test.pp - WARNING: space needed on right side of opening bracket on line 5
test.pp - WARNING: space needed on left side of opening bracket on line 5
```

**Comment by daianamezdrea** [comment]
Hi @rothsa ! Are you still working on this PR?

**Comment by rothsa** [comment]
I’m not, would you like me to close?

**Comment by DavidS** [comment]
Since rothsa is not pursuing implementing this any further and the code as is is complaining about adequate code fragments, I'll close this for now.
